src/hotspot/share/classfile/modules.cpp

```diff
@@ -559,6 +560,46 @@ void Modules::verify_archived_modules() {
   ModuleEntry::verify_archived_module_entries();
 }
 
+#if INCLUDE_CDS_JAVA_HEAP
+char* Modules::_archived_main_module_name = nullptr;
+#endif
+
+void Modules::dump_module_name() {
+  const char* module_name = Arguments::get_property("jdk.module.main");
+  if (module_name != nullptr) {
+    _archived_main_module_name = ArchiveBuilder::current()->ro_strdup(module_name);
+  }
+  ArchivePtrMarker::mark_pointer(&_archived_main_module_name);
+}
+
+void Modules::serialize(SerializeClosure* soc) {
+  soc->do_ptr(&_archived_main_module_name);
+  if (soc->reading()) {
+    const char* runtime_main_module = Arguments::get_property("jdk.module.main");
+    log_info(cds)("_archived_main_module_name %s",
+      _archived_main_module_name != nullptr ? _archived_main_module_name : "(null)");
+    if (runtime_main_module == nullptr && _archived_main_module_name != nullptr) {
+      log_info(cds)("Module %s specified during dump time but not during runtime", _archived_main_module_name);
+      log_info(cds)("Disabling optimized module handling");
+      MetaspaceShared::disable_optimized_module_handling();
+      MetaspaceShared::disable_full_module_graph();
+    }
+    if (runtime_main_module != nullptr) {
+      if (_archived_main_module_name == nullptr) {
+        log_info(cds)("Module %s specified during runtime but not during dump time", runtime_main_module);
+        log_info(cds)("Disabling optimized module handling");
+        MetaspaceShared::disable_optimized_module_handling();
+      } else if (strcmp(runtime_main_module, _archived_main_module_name) != 0) {
+        log_info(cds)("Mismatched modules: runtime %s dump time %s", runtime_main_module, _archived_main_module_name);
+        log_info(cds)("Disabling optimized module handling");
+        MetaspaceShared::disable_optimized_module_handling();
+      }
+    }
```

Comment:
I think this code can be made less verbose: 
```
    bool disable = false;
    if (runtime_main_module == nullptr) {
      if (_archived_main_module_name != nullptr) {
        log_info(cds)("Module %s specified during dump time but not during runtime", _archived_main_module_name);
        disable = true;
      }
    } else {
      if (_archived_main_module_name == nullptr) {
        log_info(cds)("Module %s specified during runtime but not during dump time", runtime_main_module);
        disable = true;
      } else if (strcmp(runtime_main_module, _archived_main_module_name) != 0) {
        log_info(cds)("Mismatched modules: runtime %s dump time %s", runtime_main_module, _archived_main_module_name);
        disable = true;
      }
    }

    if (disable) {
      log_info(cds)("Disabling optimized module handling");
      MetaspaceShared::disable_optimized_module_handling();
    }
```

Reply:
I've pushed a commit with the above simplification.